apis/python/src/tiledbsoma/_fastercsx.py

```diff
@@ -0,0 +1,270 @@
+
+        Table names must conform to the standard SOMA format, i.e., have three columns, named
+        ``soma_dim_0``, ``soma_dim_1`` and ``soma_data``. All arrays in each table column must
+        contain the same chunk count/size, and the dimension columns must be int64.
+        """
+        tables = tables if isinstance(tables, collections.abc.Sequence) else (tables,)
+
+        def chunks(a: pa.Array | pa.ChunkedArray) -> List[pa.Array]:
+            return (
+                list(a) if isinstance(a, pa.Array) else cast(List[pa.Array], a.chunks)
+            )
+
+        i = tuple(a.to_numpy() for tbl in tables for a in chunks(tbl["soma_dim_0"]))
+        j = tuple(a.to_numpy() for tbl in tables for a in chunks(tbl["soma_dim_1"]))
+        d = tuple(a.to_numpy() for tbl in tables for a in chunks(tbl["soma_data"]))
+        return CompressedMatrix.from_ijd(i, j, d, shape, format, make_sorted, context)
+
+    @property
+    def nnz(self) -> int:
+        return len(self.indices)
+
+    @property
+    def nbytes(self) -> int:
+        return int(self.indptr.nbytes + self.indices.nbytes + self.data.nbytes)
+
+    @property
+    def dtype(self) -> npt.DTypeLike:
+        return self.data.dtype
+
+    def to_scipy(
+        self, index: slice | None = None
+    ) -> scipy.sparse.csr_matrix | scipy.sparse.csc_matrix:
+        """Extract a slice on the compressed dimension and return as a dense
+        :class:`scipy.sparse.csr_matrix` or
+        :class:`scipy.sparse.csc_matrix`.
+
+        Optionally allows slicing on compressed dimension during conversion, in which case
+        an extra memory copy is avoided.
+        """
+        index = index or slice(None)
+        assert isinstance(index, slice)
+        assert index.step in (1, None)
+        idx_start, idx_end, _ = index.indices(self.indptr.shape[0] - 1)
+        n_major = max(idx_end - idx_start, 0)
+
+        if n_major == self.indptr.shape[0] - 1:
+            assert idx_start == 0 and n_major == idx_end
+            indptr = self.indptr
+            indices = self.indices
+            data = self.data
+
+        elif n_major == 0:
+            return (
+                scipy.sparse.csr_matrix((0, self.shape[1]), dtype=self.dtype)
+                if self.format == "csr"
+                else scipy.sparse.csc_matrix((self.shape[0], 0), dtype=self.dtype)
+            )
+
+        else:
+            indptr = (self.indptr[idx_start : idx_end + 1]).copy()
+            indices = self.indices[indptr[0] : indptr[-1]].copy()
+            data = self.data[indptr[0] : indptr[-1]].copy()
+            indptr -= indptr[0]  # type: ignore[misc]
+
+        shape = (
+            (n_major, self.shape[1])
+            if self.format == "csr"
+            else (self.shape[0], n_major)
+        )
+        return CompressedMatrix._to_scipy(
+            indptr, indices, data, shape, self.format, self.sorted
+        )
+
+    def to_numpy(self, index: slice | None = None) -> NDArrayNumber:
```

Comment:
I don't think is a necessary change, but we have been more consistently using `Optional` from the `typing` module

```suggestion
    def to_numpy(self, index: Optional[slice] = None) -> NDArrayNumber:
```

Reply:
Since Python 3.10, and in our style guide (Google Python style), the `type | None` is recommend, and `Optional` is not recommended.  We have a bunch of the legacy stuff from the Python 3.8 days, but need to clean it up.  Python 3.9 supports this syntax as long as you `from __futures__ import annotations`.

Reply:
IDK about policy. In our codebase we've been using `Optional` for the longest time, and i've been using `Optional` myself, certainly.

Thanks for the information that "Since Python 3.10, and in our style guide (Google Python style), the type | None is recommend, and Optional is not recommended" -- this comes as news to me. (And embarrassingly so -- the shortcoming is my own for not having been on top of that datapoint -- ideally I'd have made myself aware of this current recommendation in the Google Python style.) Re "clean it up", thank you for sharing your perspective that our status quo is unclean. (Personally I dislike "clean" in this context; I prefer to use terminology indicating that we're updating to match more recent recommendations, as such phrasing is more inclusive and value-neutral.)

Let's keep this PR as-is in this regard.

And I'll file a follow-up / lower-pri task to make a blanket pass through our codebase to convert the `Optional` typing to match more recent recommendations.

Reply:
https://github.com/single-cell-data/TileDB-SOMA/issues/3334

Reply:
I've stopped using Optional for quite some time - just out of sync with the rest of the team I guess.  Definitely advocate for migrating to a more modern syntax, although I don't see any particular urgency for the chore...

Reply:
@bkmartinjr totally agreed re non-urgency. Thank you for leading the charge here -- it's the rest of the team that's out of sync with _you_. ;)

Reply:
OIC. Thanks for the heads up. I'll start moving to using `| None` then instead of `Optional`.